### PR TITLE
Support `$inc` field `update` operator

### LINKED ISSUE #520: Support `$inc` field `update` operator
Support it for all types and edge cases. Support it with dot notation.

### PR DESCRIPTION
**Pull request by w84thesun**
Closes #520.

### CODE REVIEW

**Comment by w84thesun** [comment]
> Draft state, tests fail. I assume it is not ready for review

I should set PR reviewers when it's ready?

**Comment by seeforschauer** [comment]
> > Draft state, tests fail. I assume it is not ready for review
> 
> I should set PR reviewers when it's ready?

Agree, bcz when I check pull request review even if the state is draft I see it.

**Comment by w84thesun** [comment]
I tried to implement dot notation support for `$inc` operator. It seems that it require a lot of changes to several structures including Document and Array. Maybe it's better to do it in separate PR? @AlekSi WDYT?

**Review by seeforschauer — APPROVED** [review]

File: internal/handlers/common/error.go [hunk 1]
```diff
@@ -157,7 +170,92 @@ func formatBitwiseOperatorErr(err error, operator string, maskValue any) error {
...
+	d := must.NotFail(types.NewDocument(
+		"ok", float64(1),
+		"writeErrors", errs,
+	))
+	return d
+}
+
+// WriteError represents protocol write error.
+type WriteError struct {
+	index int64
```
Comment (seeforschauer) [inline internal/handlers/common/error.go:218]: can you please add a comment what does `index` field?

File: internal/handlers/common/error.go [hunk 2]
```diff
@@ -157,7 +170,92 @@ func formatBitwiseOperatorErr(err error, operator string, maskValue any) error {
...
+	code  ErrorCode
+	err   error
+}
+
+// Document returns wire protocol error document.
+func (we WriteError) Document() *types.Document {
+	d := must.NotFail(types.NewDocument(
+		"index", we.index,
```
Comment (seeforschauer) [inline internal/handlers/common/error.go:226]: it seems like we can have `index` different from the index of the element of the array, and that is the question for like 218.

File: integration/update_test.go [hunk 3]
```diff
@@ -72,3 +72,199 @@ func TestUpdateUpsert(t *testing.T) {
...
+			result: bson.D{{"_id", "int32"}, {"value", int64(43)}},
+		},
+
+		"FieldNotExist": {
+			filter: bson.D{{"_id", "int32"}},
+			update: bson.D{{"$inc", bson.D{{"foo", int32(1)}}}},
+			result: bson.D{{"_id", "int32"}, {"value", int32(42)}, {"foo", int32(1)}},
+		},
+
+		//"DotNotationInt": {
```
Comment (seeforschauer) [inline integration/update_test.go:232]: may we remove tests as they are not used? 
maybe it's better to create a list of what is necessary to implement for `$inc` in a separate `md` document (and suggest tests there)?

File: internal/handlers/common/params.go [hunk 4]
```diff
@@ -183,3 +185,43 @@ func parseTypeCode(alias string) (typeCode, error) {
...
+			return v2 + float64(v1), nil
+		case int32:
+			return v1 + int64(v2), nil
+		case int64:
+			return v1 + v2, nil
+		default:
+			return nil, errBadRightOperandType
+		}
+	default:
+		return nil, errBadLeftOperandType
```
Comment (seeforschauer) [inline internal/handlers/common/params.go:225]: can it be rewritten in generics?

File: internal/handlers/common/error.go (hunk 2 above)
Comment (w84thesun) [reply internal/handlers/common/error.go:226]: Sorry, 218?

File: internal/handlers/common/params.go (hunk 4 above)
Comment (w84thesun) [reply internal/handlers/common/params.go:225]: I consider to left it as is right now.

File: internal/handlers/common/error.go (hunk 1 above)
Comment (w84thesun) [reply internal/handlers/common/error.go:219]: Added

File: integration/update_test.go (hunk 3 above)
Comment (w84thesun) [reply integration/update_test.go]: Removed commented tests.

File: internal/handlers/common/error.go (hunk 2 above)
Comment (seeforschauer) [reply internal/handlers/common/error.go:227]: 218 loc (line of code) where:

> // WriteError represents protocol write error.
> type WriteError struct {
>  index int64

File: internal/handlers/common/error.go (hunk 1 above)
Comment (seeforschauer) [reply internal/handlers/common/error.go:219]: ok, thank you, and would you be so kind to make a reference where to find the index and corresponding operation description? Thank you.

File: internal/handlers/common/error.go (hunk 2 above)
Comment (seeforschauer) [reply internal/handlers/common/error.go:227]: that's was just a note for https://github.com/FerretDB/FerretDB/pull/596/files#r874474258

File: internal/handlers/common/params.go (hunk 4 above)
Comment (seeforschauer) [reply internal/handlers/common/params.go:225]: ok

File: internal/handlers/common/error.go (hunk 1 above)
Comment (w84thesun) [reply internal/handlers/common/error.go:219]: I think I can remove this field as it's not used right now.

**Review by seeforschauer — DISMISSED** [review]

File: internal/handlers/common/error.go [hunk 5]
```diff
@@ -108,11 +108,19 @@ func (e *Error) Document() *types.Document {
 	return d
 }
 
+// ProtoErr represents protocol error type.
+type ProtoErr interface {
```
Comment (AlekSi) [inline internal/handlers/common/error.go:112]: Why do we define a common interface between two implementations?

File: internal/handlers/common/error.go [hunk 6]
```diff
@@ -108,11 +108,19 @@ func (e *Error) Document() *types.Document {
 	return d
 }
 
+// ProtoErr represents protocol error type.
```
Comment (AlekSi) [inline internal/handlers/common/error.go:111]: It is not clear how that is different from `Error` type.

File: internal/handlers/common/error.go [hunk 7]
```diff
@@ -108,11 +108,19 @@ func (e *Error) Document() *types.Document {
...
+type ProtoErr interface {
+	Error() string
+	Code() ErrorCode
+	Document() *types.Document
```
Comment (AlekSi) [inline internal/handlers/common/error.go:115]: Methods should be documented

File: internal/handlers/common/error.go [hunk 8]
```diff
@@ -108,11 +108,19 @@ func (e *Error) Document() *types.Document {
...
+}
+
 // ProtocolError converts any error to wire protocol error.
 //
 // Nil panics, *Error (possibly wrapped) is returned unwrapped with true,
+// *WriteErrors is returned unwrapped with true,
```
Comment (AlekSi) [inline internal/handlers/common/error.go:121]:
```suggestion
// Nil panics, *Error or *WriteError (possibly wrapped) is returned unwrapped with true,
```

File: internal/handlers/common/error.go [hunk 9]
```diff
@@ -158,7 +171,89 @@ func formatBitwiseOperatorErr(err error, operator string, maskValue any) error {
...
+func (we *WriteErrors) Error() string {
+	var err string
+	for _, e := range *we {
+		err += e.Error() + ","
+	}
+
+	return err
+}
+
+// Code returns error code.
```
Comment (AlekSi) [inline internal/handlers/common/error.go:187]: Once methods are documented on the interface level, that comment could be just `Code implements XXX interface.`

File: internal/handlers/common/error.go [hunk 10]
```diff
@@ -158,7 +171,89 @@ func formatBitwiseOperatorErr(err error, operator string, maskValue any) error {
...
+// Unwrap implements standard error unwrapping interface.
+func (we *WriteErrors) Unwrap() error {
+	for _, e := range *we {
+		return &e
+	}
+	return nil
+}
+
+// Document returns wire protocol error document.
+func (we WriteErrors) Document() *types.Document {
```
Comment (AlekSi) [inline internal/handlers/common/error.go:204]: Why other methods are implemented on the pointer, but that one is one the value?

File: internal/handlers/common/error.go [hunk 11]
```diff
@@ -158,7 +171,89 @@ func formatBitwiseOperatorErr(err error, operator string, maskValue any) error {
...
+	}
+
+	d := must.NotFail(types.NewDocument(
+		"ok", float64(1),
+		"writeErrors", errs,
+	))
+	return d
+}
+
+// WriteError represents protocol write error.
```
Comment (AlekSi) [inline internal/handlers/common/error.go:217]: So is ProtoErr. We need better descriptions

File: internal/handlers/common/error.go [hunk 12]
```diff
@@ -158,7 +171,89 @@ func formatBitwiseOperatorErr(err error, operator string, maskValue any) error {
...
+type WriteError struct {
+	code ErrorCode
+	err  error
+}
+
+// Document returns wire protocol error document.
+func (we WriteError) Document() *types.Document {
```
Comment (AlekSi) [inline internal/handlers/common/error.go:224]: Any particular reason why methods order is different?

File: internal/handlers/common/params.go [hunk 13]
```diff
@@ -213,3 +215,43 @@ func parseTypeCode(alias string) (typeCode, error) {
 
 	return code, nil
 }
+
+func addNumbers(v1, v2 any) (any, error) {
```
Comment (AlekSi) [inline internal/handlers/common/params.go:219]: Function that accepts anything and returns anything _must_ be documented

File: internal/handlers/common/update.go [hunk 14]
```diff
@@ -41,6 +42,66 @@ func UpdateDocument(doc, update *types.Document) error {
 				}
 			}
 
+		case "$inc":
+			incDoc, err := AssertType[*types.Document](updateV)
```
Comment (AlekSi) [inline internal/handlers/common/update.go:46]: `AssertType` documentation says:
> If a custom error is needed, use a normal Go type assertion instead [...]

File: internal/handlers/common/params.go [hunk 15]
```diff
@@ -213,3 +215,43 @@ func parseTypeCode(alias string) (typeCode, error) {
...
+	switch v1 := v1.(type) {
+	case float64:
+		switch v2 := v2.(type) {
+		case float64:
+			return v1 + v2, nil
```
Comment (AlekSi) [inline internal/handlers/common/params.go:224]: That code would happily sum NaN and infinity, but there are no tests for that

**Review by AlekSi — CHANGES_REQUESTED** [review]

File: internal/handlers/common/params.go (hunk 15 above)
Comment (w84thesun) [reply internal/handlers/common/params.go:224]: I've added a test for that case.

File: internal/handlers/common/update.go (hunk 14 above)
Comment (w84thesun) [reply internal/handlers/common/update.go]: Fixed

File: internal/handlers/common/error.go (hunk 6 above)
Comment (w84thesun) [reply internal/handlers/common/error.go]: In fact, `Error` type should be `CommandError`. 
Whereas `ProtoErr` represents a common protocol error type (it could be CommandError or WriteErrors).

File: internal/handlers/common/error.go (hunk 7 above)
Comment (w84thesun) [reply internal/handlers/common/error.go]: Added comments

File: integration/update_test.go [hunk 16]
```diff
@@ -72,3 +73,186 @@ func TestUpdateUpsert(t *testing.T) {
...
 	AssertEqualDocuments(t, bson.D{{"_id", id}, {"foo", "qux"}}, doc)
 }
+
+func TestUpdateIncOperatorErrors(t *testing.T) {
+	t.Parallel()
+
+	for name, tc := range map[string]struct {
+		filter bson.D
+		update bson.D
+		err    *mongo.WriteError
```
Comment (seeforschauer) [inline integration/update_test.go:83]: why not create bare value, not a pointer? 
Mongo driver returns not a pointer but a bare struct.

Comment (seeforschauer) [reply integration/update_test.go:83]: ah I see now - to make nil pointer check

File: integration/update_test.go [hunk 17]
```diff
@@ -72,3 +73,186 @@ func TestUpdateUpsert(t *testing.T) {
...
+			_, err := collection.UpdateOne(ctx, tc.filter, tc.update)
+			if tc.err != nil {
+				AssertEqualWriteError(t, tc.err, err)
+				return
+			}
+			require.NoError(t, err)
+
+			var actual bson.D
+			err = collection.FindOne(ctx, tc.filter).Decode(&actual)
+			require.NoError(t, err)
```
Comment (seeforschauer) [inline integration/update_test.go:141]: must not reach here, it's a error test.

File: integration/update_test.go [hunk 18]
```diff
@@ -72,3 +73,186 @@ func TestUpdateUpsert(t *testing.T) {
...
+			err: &mongo.WriteError{
+				Code:    14,
+				Message: `Cannot apply $inc to a value of non-numeric type. {_id: "string"} has the field 'value' of non-numeric type string`,
+			},
+		},
+	} {
+		name, tc := name, tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			ctx, collection := setup(t, shareddata.Scalars)
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply integration/update_test.go:130]: We need separate collection names to ensure update operations are not executed on the same data.

File: integration/update_test.go (hunk 17 above)
Comment (w84thesun) [reply integration/update_test.go]: Nice catch, thanks!

File: integration/update_test.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (seeforschauer) [reply integration/update_test.go:130]: In the `setup` there is an `InsertOne`. 
How does `insert` work? Does it add or replace? I assume, it adds but `_id`s are the same. 
If it adds then how it is still working. Are we trying to update all existing `_id`s? 
What happens there?

File: integration/helpers.go [hunk 19]
```diff
@@ -184,3 +184,22 @@ func CollectKeys(t testing.TB, doc bson.D) []string {
 
 	return res
 }
+
+// AssertEqualWriteError compares expected mongo.WriteError Message and Code with actual error.
+func AssertEqualWriteError(t *testing.T, expected *mongo.WriteError, actual error) bool {
```
Comment (seeforschauer) [inline integration/helpers.go:189]: I'm not sure, but it also can shadow the MongoDB errors checking for example when we expect in tests a certain error type, but MongoDB gives another type, it will bypass.

File: internal/handlers/common/error.go [hunk 20]
```diff
@@ -108,21 +139,87 @@ func (e *Error) Document() *types.Document {
...
+}
+
+// Unwrap implements standard error unwrapping interface.
+func (we *WriteError) Unwrap() error {
+	return we.err
+}
+
+// Document implements ProtoErr interface.
+// Fields "code" and "errmsg" must always be filled in so that clients can parse the error message.
+func (we *WriteError) Document() *types.Document {
```
Comment (seeforschauer) [inline internal/handlers/common/error.go:217]: Could you please add a comment that otherwise it would parse it as a command error?

File: integration/helpers.go (hunk 19 above)
Comment (seeforschauer) [reply integration/helpers.go:189]: No shadowing here - it can be seen in deep debug on mongo driver.

File: internal/handlers/common/error.go [hunk 21]
```diff
@@ -108,21 +139,88 @@ func (e *Error) Document() *types.Document {
...
-	return NewError(errInternalError, err).(*Error), false
+// Unwrap implements standard error unwrapping interface.
+func (we *WriteErrors) Unwrap() error {
+	for _, e := range *we {
+		return &e
+	}
+	return nil
+}
+
+// Document implements ProtoErr interface..
```
Comment (seeforschauer) [inline internal/handlers/common/error.go:179]: Could you please add here some more from your debugging adventures?

File: integration/update_test.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (seeforschauer) [reply integration/update_test.go:130]: each time a new collection, ok.

File: internal/handlers/common/error.go (hunk 20 above)
Comment (w84thesun) [reply internal/handlers/common/error.go:220]: Added.

File: internal/handlers/common/error.go (hunk 21 above)
Comment (w84thesun) [reply internal/handlers/common/error.go]: Added.

**Review by rumyantseva — COMMENTED** [review]
I still miss some context to provide a good code review here. So, I mostly ask questions, some of them might be completely irrelevant.

Overall, it seems that such choices (which types can be `$inc` and which cannot) need to be documented. I guess, it's a separate task to document those things better?

File: internal/handlers/common/error.go [hunk 22]
```diff
@@ -161,4 +261,6 @@ func formatBitwiseOperatorErr(err error, operator string, maskValue any) error {
 // check interfaces
 var (
 	_ error = (*Error)(nil)
+	_ error = (*WriteError)(nil)
```
Comment (AlekSi) [inline internal/handlers/common/error.go:264]: Is there an API that could return a single write error of this type? Should both `WriteError` and `WriteErrors` implement error interface?

And we should check that Error/CommandError and WriteErrors implement ProtoErr interface

File: internal/handlers/common/error.go (hunk 5 above)
Comment (AlekSi) [reply internal/handlers/common/error.go]: That was a question about ProtoErr interface definition in the source ode file – it was sandwiched between Error and WriteErrors definitions. That's not the case now

File: integration/update_test.go [hunk 23]
```diff
@@ -72,3 +73,178 @@ func TestUpdateUpsert(t *testing.T) {
...
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply integration/update_test.go]: Added tests for documents and arrays.

File: integration/update_test.go [hunk 24]
```diff
@@ -72,3 +73,178 @@ func TestUpdateUpsert(t *testing.T) {
...
+		"FieldNotExist": {
+			filter: bson.D{{"_id", "int32"}},
+			update: bson.D{{"$inc", bson.D{{"foo", int32(1)}}}},
+			result: bson.D{{"_id", "int32"}, {"value", int32(42)}, {"foo", int32(1)}},
+		},
+	} {
+		name, tc := name, tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			ctx, collection := setup(t, shareddata.Scalars, shareddata.Composites)
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply integration/update_test.go]: Fixed.

File: internal/handlers/common/error.go [hunk 25]
```diff
@@ -51,7 +51,38 @@ const (
 	ErrRegexMissingParen = ErrorCode(51091) // Location51091
 )
 
-// Error represents wire protocol error.
+// ProtoErr represents protocol error type.
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply internal/handlers/common/error.go:54]: We need `WriteErrors` to return correct errors to clients. In order to use CommandError and WriteError at `conn.go` we introduce this interface.

File: internal/handlers/common/error.go (hunk 11 above)
Comment (w84thesun) [reply internal/handlers/common/error.go:195]: The second sentence explains why it exists.

File: internal/handlers/common/params.go [hunk 26]
```diff
@@ -105,10 +105,12 @@ func AssertType[T types.Type](value any) (T, error) {
...
 var (
-	errUnexpectedType = fmt.Errorf("unexpected type")
-	errNotWholeNumber = fmt.Errorf("not a whole number")
-	errNegativeNumber = fmt.Errorf("negative number")
-	errNotBinaryMask  = fmt.Errorf("not a binary mask")
+	errUnexpectedType      = fmt.Errorf("unexpected type")
+	errNotWholeNumber      = fmt.Errorf("not a whole number")
+	errNegativeNumber      = fmt.Errorf("negative number")
+	errNotBinaryMask       = fmt.Errorf("not a binary mask")
+	errBadLeftOperandType  = fmt.Errorf("bad left operand type")
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply internal/handlers/common/params.go]: > Is a similar wording used somewhere else, so this "bad" is consistent with it?

Nope, I don't think so.
Changed to `errUnexpectedLeftOpType`.

File: integration/update_test.go [hunk 27]
```diff
@@ -72,3 +73,178 @@ func TestUpdateUpsert(t *testing.T) {
 	require.NoError(t, err)
 	AssertEqualDocuments(t, bson.D{{"_id", id}, {"foo", "qux"}}, doc)
 }
+
+func TestUpdateIncOperatorErrors(t *testing.T) {
+	t.Parallel()
+
+	for name, tc := range map[string]struct {
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply integration/update_test.go:80]: Added tests for documents and arrays.

File: integration/update_test.go [hunk 28]
```diff
@@ -72,3 +73,178 @@ func TestUpdateUpsert(t *testing.T) {
 	require.NoError(t, err)
 	AssertEqualDocuments(t, bson.D{{"_id", id}, {"foo", "qux"}}, doc)
 }
+
+func TestUpdateIncOperatorErrors(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply integration/update_test.go:77]: I thought that the test name reflects this by itself. Like we have `TestUpdateIncOperatorErrors` and `TestUpdateIncOperator`. The first one checks errors and the second one checks responses.

File: integration/helpers.go [hunk 29]
```diff
@@ -184,3 +184,22 @@ func CollectKeys(t testing.TB, doc bson.D) []string {
...
+		return assert.Equal(t, expected, actual)
+	}
+
+	if len(writeException.WriteErrors) != 1 {
+		return assert.Equal(t, expected, actual)
+	}
+
+	actualWriteErr := writeException.WriteErrors[0]
+
+	return assert.Equal(t, expected.Message, actualWriteErr.Message) &&
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply integration/helpers.go:203]: `assert.Equal` would print message text or code if it's not equal. The potential problem here is that if the code and message are not equal it will print a message or a code not both of them, so you would require 2 test runs to fix it.

File: internal/handlers/common/error.go (hunk 20 above)
Comment (seeforschauer) [reply internal/handlers/common/error.go:223]: Thank you.

> Added.

File: internal/handlers/common/error.go (hunk 22 above)
Comment (w84thesun) [reply internal/handlers/common/error.go]: No, `WriteError` always returned as a part of `WriteErrors`. I made `WriteError` unexported.

File: integration/update_test.go [hunk 30]
```diff
@@ -72,3 +73,194 @@ func TestUpdateUpsert(t *testing.T) {
...
+			err: &mongo.WriteError{
+				Code:    14,
+				Message: `Cannot apply $inc to a value of non-numeric type. {_id: "string"} has the field 'value' of non-numeric type string`,
+			},
+		},
+	} {
+		name, tc := name, tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			ctx, collection := setup(t, shareddata.Scalars, shareddata.Composites)
```
Comment (seeforschauer) [inline integration/update_test.go:147]: Would you consider adding here only document, string, and array? 
It requires each time a lot of connections / less unused data is better.

File: integration/update_test.go [hunk 31]
```diff
@@ -72,3 +73,194 @@ func TestUpdateUpsert(t *testing.T) {
...
+		"FieldNotExist": {
+			filter: bson.D{{"_id", "int32"}},
+			update: bson.D{{"$inc", bson.D{{"foo", int32(1)}}}},
+			result: bson.D{{"_id", "int32"}, {"value", int32(42)}, {"foo", int32(1)}},
+		},
+	} {
+		name, tc := name, tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			ctx, collection := setup(t, shareddata.Scalars)
```
Comment (seeforschauer) [inline integration/update_test.go:254]: I would suggest the same here - double, double-nan, int34, and int64 are much less than shareddata.Scalars.

File: internal/handlers/common/error.go [hunk 32]
```diff
@@ -108,21 +142,69 @@ func (e *Error) Document() *types.Document {
...
+// Unwrap implements standard error unwrapping interface.
+func (we *WriteErrors) Unwrap() error {
+	for _, e := range *we {
+		return errors.New(e.err)
+	}
+	return nil
+}
+
+// Document implements ProtoErr interface.
+// "writeErrors" field must present in result document so that clients could parse it as WriteErrors.
```
Comment (seeforschauer) [inline internal/handlers/common/error.go:184]:
```suggestion
//  "writeErrors" field must be present in the result document so that clients can parse it as WriteErrors.
```

File: internal/handlers/common/error.go [hunk 33]
```diff
@@ -108,21 +142,69 @@ func (e *Error) Document() *types.Document {
...
+// Fields "code" and "errmsg" must always be filled in so that clients can parse the error message,
+// otherwise mongo client would parse it as a CommandError.
```
Comment (seeforschauer) [inline internal/handlers/common/error.go:186]:
```suggestion
// Otherwise, the mongo client would parse it as a CommandError.
```

File: internal/handlers/common/error.go [hunk 34]
```diff
@@ -108,21 +142,69 @@ func (e *Error) Document() *types.Document {
...
+func (we *WriteErrors) Unwrap() error {
+	for _, e := range *we {
+		return errors.New(e.err)
+	}
+	return nil
+}
+
+// Document implements ProtoErr interface.
+// "writeErrors" field must present in result document so that clients could parse it as WriteErrors.
+// Fields "code" and "errmsg" must always be filled in so that clients can parse the error message,
```
Comment (seeforschauer) [inline internal/handlers/common/error.go:185]:
```suggestion
// Fields "code" and "errmsg" must always be filled in so that clients can parse the error message.
```

File: internal/handlers/common/error.go [hunk 35]
```diff
@@ -108,21 +142,69 @@ func (e *Error) Document() *types.Document {
...
+// Code implements ProtoErr interface.
+func (we *WriteErrors) Code() ErrorCode {
+	for _, e := range *we {
+		return e.code
 	}
+	return errUnset
+}
 
-	return NewError(errInternalError, err).(*Error), false
+// Unwrap implements standard error unwrapping interface.
```
Comment (seeforschauer) [inline internal/handlers/common/error.go:175]:
```suggestion
// Unwrap implements a standard error unwrapping interface.
```

File: internal/handlers/common/error.go [hunk 36]
```diff
@@ -108,21 +142,69 @@ func (e *Error) Document() *types.Document {
...
-// Nil panics, *Error (possibly wrapped) is returned unwrapped with true,
-// any other value is wrapped with InternalError and returned with false.
-func ProtocolError(err error) (*Error, bool) {
-	if err == nil {
-		panic("err is nil")
+// WriteErrors represents slice of protocol write errors.
+// It could be returned for Update, Insert, Delete, Replace operations.
+type WriteErrors []writeError
+
+// NewWriteErrorMsg creates new protocol write error with given ErrorCode and message.
```
Comment (seeforschauer) [inline internal/handlers/common/error.go:149]:
```suggestion
// NewWriteErrorMsg creates a new protocol write error with given ErrorCode and message.
```

File: internal/handlers/common/error.go [hunk 37]
```diff
@@ -108,21 +142,69 @@ func (e *Error) Document() *types.Document {
...
 
-// ProtocolError converts any error to wire protocol error.
-//
-// Nil panics, *Error (possibly wrapped) is returned unwrapped with true,
-// any other value is wrapped with InternalError and returned with false.
-func ProtocolError(err error) (*Error, bool) {
-	if err == nil {
-		panic("err is nil")
+// WriteErrors represents slice of protocol write errors.
+// It could be returned for Update, Insert, Delete, Replace operations.
```
Comment (seeforschauer) [inline internal/handlers/common/error.go:146]:
```suggestion
// It could be returned for Update, Insert, Delete, and Replace operations.
```

File: internal/handlers/common/error.go [hunk 38]
```diff
@@ -108,21 +142,69 @@ func (e *Error) Document() *types.Document {
...
 }
 
-// ProtocolError converts any error to wire protocol error.
-//
-// Nil panics, *Error (possibly wrapped) is returned unwrapped with true,
-// any other value is wrapped with InternalError and returned with false.
-func ProtocolError(err error) (*Error, bool) {
-	if err == nil {
-		panic("err is nil")
+// WriteErrors represents slice of protocol write errors.
```
Comment (seeforschauer) [inline internal/handlers/common/error.go:145]:
```suggestion
// WriteErrors represents a slice of protocol write errors.
```

File: internal/handlers/common/error.go [hunk 39]
```diff
@@ -51,7 +51,41 @@ const (
...
+		return writeErr, true
+	}
+
+	return NewError(errInternalError, err).(*Error), false
+}
+
+// CommandError represents wire protocol command error.
+type CommandError = Error
+
+// Error is a deprecated name for CommandError; use the later in the new code instead.
```
Comment (seeforschauer) [inline internal/handlers/common/error.go:88]:
```suggestion
// Error is a deprecated name for CommandError; instead, use the later version in the new code.
```

File: internal/handlers/common/params.go [hunk 40]
```diff
@@ -213,3 +215,46 @@ func parseTypeCode(alias string) (typeCode, error) {
 
 	return code, nil
 }
+
+// addNumbers returns the result of v1 and v2 addition and error if addition failed.
+// The v1 and v2 parameters could be float64, int32, int64.
+// The result would be the biggest type possible, ie int32 + int64 produce int64 result.
```
Comment (seeforschauer) [inline internal/handlers/common/params.go:221]:
```suggestion
// The result would be the broader type possible, i.e. int32 + int64 produces int64.
```

File: integration/update_test.go (hunk 30 above)
Comment (w84thesun) [reply integration/update_test.go:147]: Scalars + composites are 49 documents in total. I don't think there should be any connection problem inserting 49 documents with one InsertMany request. Otherwise, I should extract those documents from sets or write my own for each test which makes those sets kinda useless.

File: integration/update_test.go (hunk 31 above)
Comment (w84thesun) [reply integration/update_test.go:254]: Same as above.

Comment (seeforschauer) [reply integration/update_test.go:254]: I'm sorry I just saw in my own PR, where a lot of data was not necessary and it was taken too long for tests to retest.
And, my second argument, is that also we'd probably get rid someday of shareddata, it's in a comment of shared-data.

Comment (w84thesun) [reply integration/update_test.go]: Updated.

**Review by seeforschauer — APPROVED** [review]
LGTM.

**Review by AlekSi — DISMISSED** [review]

**Comment by AlekSi** [comment]
@rumyantseva PTAL and merge with the right commit body if it looks good to you

File: internal/handlers/common/error.go [hunk 41]
```diff
@@ -108,21 +142,69 @@ func (e *Error) Document() *types.Document {
...
+	for _, e := range *we {
+		return errors.New(e.err)
+	}
+	return nil
+}
+
+// Document implements ProtoErr interface.
+//  "writeErrors" field must be present in the result document so that clients can parse it as WriteErrors.
+// Fields "code" and "errmsg" must always be filled in so that clients can parse the error message.
+// Otherwise, the mongo client would parse it as a CommandError.
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply internal/handlers/common/error.go]: Fixed

File: internal/handlers/common/error.go [hunk 42]
```diff
@@ -160,5 +242,8 @@ func formatBitwiseOperatorErr(err error, operator string, maskValue any) error {
 
 // check interfaces
 var (
-	_ error = (*Error)(nil)
+	_ error    = (*Error)(nil)
+	_ error    = (*WriteErrors)(nil)
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply internal/handlers/common/error.go]: Fixed

File: internal/handlers/common/update.go [hunk 43]
```diff
@@ -41,6 +42,66 @@ func UpdateDocument(doc, update *types.Document) error {
 				}
 			}
 
+		case "$inc":
+			incDoc, ok := updateV.(*types.Document)
```
Comment (rumyantseva) [inline internal/handlers/common/update.go:46]: Why don't we move that long construction to a separate function? So then it's easier to test and the cyclomatic complexity is a bit lower (and it's easier to read it). I don't want to block this PR merge by it, but I believe we should discuss this topic.

Comment (w84thesun) [reply internal/handlers/common/update.go:46]: There would be no test impact if I move it to a separate function as we test with integration tests mostly.
I believe it's just because we don't have an agreement on when it's time to extract a function right now.

**Review by rumyantseva — APPROVED** [review]
LGTM. I have a question about long switch-case constructions, but as they are presented in various places, not only in this PR, I will raise it separately.
